### PR TITLE
Add tests for %FunctionPrototype% restricted properties

### LINKED ISSUE #211: Tests related to ES5 step 13.2.19 are not valid in ES6
Per https://people.mozilla.org/~jorendorff/es6-draft.html#sec-forbidden-extensions --- for strict functions, and for functions created by new syntactic forms (method definitions, generators, arrow functions, class constructors), no "own" properties "caller" or "arguments" are created.

These properties are defined as accessors which throw, only on the intrinsic %FunctionPrototype% object, with `PropertyDescriptor {[[Get]]: thrower, [[Set]]: thrower, [[Enumerable]]: false, [[Configurable]]: true}).`

Tests relating to this should probably live in a new directory specific to "forbidden extensions" under "language", and should be updated to reflect the current state of the spec

### PR DESCRIPTION
**Pull request by caitp**
First stab at updating the outdated tests

Things missing:
  - asserting that new syntactic forms inherit from %FunctionPrototype% (should be covered elsewhere, eg assertions about [25.2.3](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-properties-of-the-generatorfunction-prototype-object))
  - more meaningful tests for enumerability and configurability of the properties

Closes #211

### CODE REVIEW

File: test/language/class/definition/ClassMethod_restricted-properties.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    have own properties "caller" or "arguments", but inherit them from
+    %FunctionPrototype%.
+---*/
+
+class Class {
+  method() {}
+};
+
+assert.sameValue(Class.method.hasOwnProperty('caller'), false, 'Functions created using MethodDefinition syntactic form do not have own property "caller"');
+assert.sameValue(Class.method.hasOwnProperty('arguments'), false, 'Functions created using MethodDefinition syntactic form do not have own property "arguments"');
```
Comment (anba) [inline test/language/class/definition/ClassMethod_restricted-properties.js]: `Class.method` -> `Class.prototype.method`

File: test/built-ins/intrinsics/FunctionPrototype/FunctionPrototype_restricted-property-caller.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+/*--- 
+description: Intrinsic %FunctionPrototype% has poisoned own property "caller"
+---*/
+
+var FunctionPrototype = Function.prototype;
+
+assert.sameValue(FunctionPrototype.hasOwnProperty('caller'), true, 'The result of %FunctionPrototype%.hasOwnProperty("caller") is true');
+
+var descriptor = Object.getOwnPropertyDescriptor(FunctionPrototype, 'caller');
+assert.sameValue(descriptor.configurable, 'function', '%FunctionPrototype%.caller is a configurable property');
```
Comment (anba) [inline test/built-ins/intrinsics/FunctionPrototype/FunctionPrototype_restricted-property-caller.js]: `'function'` -> `true`

File: test/built-ins/intrinsics/FunctionPrototype/FunctionPrototype_restricted-property-arguments.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+/*--- 
+description: Intrinsic %FunctionPrototype% has poisoned own property "arguments"
+---*/
+
+var FunctionPrototype = Function.prototype;
+
+assert.sameValue(FunctionPrototype.hasOwnProperty('arguments'), true, 'The result of %FunctionPrototype%.hasOwnProperty("arguments") is true');
+
+var descriptor = Object.getOwnPropertyDescriptor(FunctionPrototype, 'arguments');
+assert.sameValue(descriptor.configurable, 'function', '%FunctionPrototype%.arguments is a configurable property');
```
Comment (anba) [inline test/built-ins/intrinsics/FunctionPrototype/FunctionPrototype_restricted-property-arguments.js]: `'function'` -> `true`

**Comment by anba** [comment]
I guess test/language/statements/function/13.1-29-s.js was erroneously deleted?

> Things missing:
>     ???

Missing coverage: getter and setter method definitions and functions created with `Function` and `GeneratorFunction`.

File: test/built-ins/intrinsics/FunctionPrototype/FunctionPrototype_restricted-property-arguments.js (hunk 3 above)
Comment (caitp) [inline test/built-ins/intrinsics/FunctionPrototype/FunctionPrototype_restricted-property-arguments.js]: oop --- there were a few of these somehow, I think my text editor was saving things wrong while I was building content_shell --- keystrokes not registering or something. Anyway, fixing those up as they get noticed

**Comment by caitp** [comment]
> Missing coverage: getter and setter method definitions and functions created with Function and GeneratorFunction.

Added these to the StrictFunction_restricted-properties.js and Generator_restricted-properties.js files

File: test/built-ins/intrinsics/FunctionPrototype/FunctionPrototype_restricted-property-arguments.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+var descriptor = Object.getOwnPropertyDescriptor(FunctionPrototype, 'arguments');
+assert.sameValue(descriptor.configurable, true, '%FunctionPrototype%.arguments is a configurable property');
+assert.sameValue(descriptor.enumerable, false, '%FunctionPrototype%.arguments is a non-enumerable property');
+assert.sameValue(typeof descriptor.get, 'function', '%FunctionPrototype%.arguments is an accessor property');
+assert.sameValue(typeof descriptor.set, 'function', '%FunctionPrototype%.arguments is an accessor property');
+assert.sameValue(descriptor.get, descriptor.set, '%FunctionPrototype%.arguments getter/setter are both %ThrowTypeError%');
+
+assert.throws(TypeError, function() {
+  return FunctionPrototype.caller;
```
Comment (arv) [inline test/built-ins/intrinsics/FunctionPrototype/FunctionPrototype_restricted-property-arguments.js]: should be arguments here and on line 24

Comment (caitp) [inline test/built-ins/intrinsics/FunctionPrototype/FunctionPrototype_restricted-property-arguments.js]: derp, good eye thanks

File: test/built-ins/intrinsics/FunctionPrototype/FunctionPrototype_restricted-property-arguments.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+/*--- 
+description: Intrinsic %FunctionPrototype% has poisoned own property "arguments"
+---*/
+
+var FunctionPrototype = Function.prototype;
+
+assert.sameValue(FunctionPrototype.hasOwnProperty('arguments'), true, 'The result of %FunctionPrototype%.hasOwnProperty("arguments") is true');
+
+var descriptor = Object.getOwnPropertyDescriptor(FunctionPrototype, 'arguments');
+assert.sameValue(descriptor.configurable, true, '%FunctionPrototype%.arguments is a configurable property');
```
Comment (arv) [inline test/built-ins/intrinsics/FunctionPrototype/FunctionPrototype_restricted-property-arguments.js]: Maybe you can use propertyHelper.js here?

Comment (caitp) [inline test/built-ins/intrinsics/FunctionPrototype/FunctionPrototype_restricted-property-arguments.js]: Done for the configurable/not enumerable tests

**Comment by caitp** [comment]
so, anything needed on this?

File: test/language/class/definition/ClassMethod_restricted-properties.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+/*--- 
+ description: >
+    Functions created using MethodDefinition syntactic form do not
+    have own properties "caller" or "arguments", but inherit them from
+    %FunctionPrototype%.
+---*/
+
+class Class {
+  method() {}
```
Comment (arv) [inline test/language/class/definition/ClassMethod_restricted-properties.js]: No test for getter/setter/constructor?

Comment (caitp) [inline test/language/class/definition/ClassMethod_restricted-properties.js]: constructor is tested in the other class-related thing. I'm not sure what the spec says about `get foo()/set foo()` for this though

Comment (caitp) [inline test/language/class/definition/ClassMethod_restricted-properties.js]: I guess `get PropertyName ( ) { FunctionBody }` and the setter variant are MethodDefinitions, so they're covered... adding tests

Comment (caitp) [inline test/language/class/definition/ClassMethod_restricted-properties.js]: done

Comment (caitp) [inline test/language/class/definition/ClassMethod_restricted-properties.js]: fixed up

**Comment by bterlson** [comment]
This seems good. Can you add an es6id, though?

**Comment by caitp** [comment]
Will do that for tomorrow

**Comment by caitp** [comment]
sorry it took so long, added es6id annotations and squashed

**Pull request merged**